examples/model_selection/plot_grid_search_digits.py

```diff
@@ -13,56 +13,185 @@
+n_samples = len(digits.images)
+X = digits.images.reshape((n_samples, -1))
+y = digits.target == 7
+print(
+    f"The number of images is {X.shape[0]} and each image contains {X.shape[1]} pixels"
+)
+
+# %%
+# As presented in the introduction, the data will be split into a training
+# and an testing sets of equal size.
+from sklearn.model_selection import train_test_split
 
-# Split the dataset in two equal parts
 X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=0)
 
-# Set the parameters by cross-validation
+# %%
+# Define our grid-search strategy
+# -------------------------------
+#
+# We will select a classifier by searching the best hyper-parameters on folds
+# of the training set. To do this, we first need to define the values of
+# the hyper-parameters and the scores to compute to later select the
+# best candidate.
+
 tuned_parameters = [
-    {"kernel": ["rbf"], "gamma": [1e-3, 1e-4], "C": [1, 10, 100, 1000]},
-    {"kernel": ["linear"], "C": [1, 10, 100, 1000]},
+    {"solver": ["newton-cg"], "penalty": ["l2", "none"], "C": [1, 10, 100]},
+    {
+        "solver": ["saga"],
+        "penalty": ["elasticnet", "l2", "none"],
+        "C": [1, 10, 100],
+    },
 ]
 
 scores = ["precision", "recall"]
 
-for score in scores:
-    print("# Tuning hyper-parameters for %s" % score)
-    print()
-
-    clf = GridSearchCV(SVC(), tuned_parameters, scoring="%s_macro" % score)
-    clf.fit(X_train, y_train)
-
-    print("Best parameters set found on development set:")
-    print()
-    print(clf.best_params_)
-    print()
-    print("Grid scores on development set:")
-    print()
-    means = clf.cv_results_["mean_test_score"]
-    stds = clf.cv_results_["std_test_score"]
-    for mean, std, params in zip(means, stds, clf.cv_results_["params"]):
-        print("%0.3f (+/-%0.03f) for %r" % (mean, std * 2, params))
-    print()
-
-    print("Detailed classification report:")
-    print()
-    print("The model is trained on the full development set.")
-    print("The scores are computed on the full evaluation set.")
-    print()
-    y_true, y_pred = y_test, clf.predict(X_test)
-    print(classification_report(y_true, y_pred))
-    print()
-
-# Note the problem is too easy: the hyperparameter plateau is too flat and the
-# output model is the same for precision and recall with ties in quality.
+# %%
+# We can also define a function that will implement the strategy to use to
+# select the best candidate from the `cv_results_` returns from the
+# :class:`~sklearn.model_selection.GridSearchCV`. Once the candidate selected,
+# it will be automatically refitted by the
+# :class:`~sklearn.model_selection.GridSearchCV` instance. Here, the strategy
+# is to short list the model which are the best in terms of precision and
+# recall. From the selected models, we finally select the fastest model at
+# predicting.
+#
+# Note that these choices are completely arbitrary.
```

Comment:
```suggestion
# We can also define a function to be passed to the `refit` parameter of the
# `GridSearchCV` instance. It will implement the custom strategy to select the
# best candidate from the `cv_results_` attribute of the
# :class:`~sklearn.model_selection.GridSearchCV`. Once the candidate is
# selected, it is automatically refitted by the
# :class:`~sklearn.model_selection.GridSearchCV` instance.
# 
# Here, the strategy is to short-list the model which are the best in terms of
# precision and recall. From the selected models, we finally select the fastest
# model at predicting. Notice that these custom choices are completely
# arbitrary.

```